### ISSUE TITLE
Make rule kind available in the rule ctx.

### ISSUE THREAD

**Issue by hsyed** [bazelbuild/bazel#5887]
The Kotlin rules pass the rule kind (--e.g., `kt_jvm_library`) on to the builder for stamping the manifest and other purposes. 

I have looked around but I could not find this in the ctx. It is available on the `rule` attribute in an aspect, but adding an aspect just to extract this would be silly. This should be available in the ctx.

**Comment by laurentlb** [comment]
You shouldn't rely on the rule kind, use providers instead to pass information.

(it should be possible for someone else to create a new rule behaving the same way as `kt_jvm_library`, so the code shouldn't take into account the rule kind).

**Comment by hsyed** [comment]
To be clear I need the rule kind for the current target in the ctx of the current target. Providers would give me the rule kind of others — unless you mean using an aspect to extract the kind and put it in provider but that would be overkill.

**Comment by hsyed** [comment]
Also the kotlin builder targets multiple platforms. It infers what to do by looking at “jvm” / “js” and it stamps the manifest. 

The common skylark codepaths could check weather the rule is binary or test. 

Not having this information means it has to passed arround in params.

**Comment by laurentlb** [comment]
I see, so it's just for convenience (you have the information because you decide which implementation function gets called).

If you don't want to have an extra argument to your functions, a workaround is to add a private attribute to your rule.

**Comment by hsyed** [comment]
Hmm I hadn't considered that, I'll use that then for now. 

I remembered why I added the `rule_kind` initially -- so here is one more argument for it. When I was prototyping strict deps and  support for the `unused_deps`tool I copied The format of the `.jar-2.params` files generated by the native java rules. The `rule_kind`. `unused_deps` does need to see the rule kind if it's going to support multiple rule sets. So, if the native java rules can use it to convey key information to other tools then maybe it should be open skylark as well.  

Feel free to close the issue.
